Improve performance in SqliteStorage::get_bagfile_size()

MichaelOrlov opened the pull request:
- Closes https://github.com/ros2/rosbag2/issues/647
- According to the https://www.sqlabs.com/blog/2010/12/getting-the-size-of-an-sqlite-database/. 
In SQLite3, the DB file size can be calculated as db `page_count * page_size`.

- Use DB query to get page_count and page_size instead of a filesystem blocking call.
- Added regression unit test to make sure that `SqliteStorage::get_bagfile_size()` returns correct value for both storages SQLite3 and MCAP.

morlov-apexai committed:
Improve performance in SqliteStorage::get_bagfile_size()

- In SQLite3 DB file size is equal to the db page_count * page_size
- Use DB query to get page_count and page_size instead of filesystem
blocking call.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Address integration tests failure

- Don't use static objects for the SqliteStatement, it causes DB lock
and failure to close and open it again.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
@emersonknapp If you have experience with SQLite3 storage especially with raw database access I would appreciate your help with this PR.

I've made an implementation that is IMO pretty trivial and simple. However, I am getting one integration test failure with an error message that the database table is locked during the call for the open() method I think. I wouldn't expect this type of error for any scenarios in the current implementation and have no more ideas why DB table became locked.  
Any ideas or thoughts will be helpful. BTW  this error doesn't reproduce on my local setup with regular colcon build/test commands.
Here is the fragment of the log with failure.
```
[ RUN      ] TestRecordEndToEnd/RecordFixture.rosbag2_record_and_play_multiple_topics_with_filter/0
stdin is not a terminal device. Keyboard handling disabled.[INFO] [1702867875.584426960] [rosbag2_recorder]: Press SPACE for pausing/resuming
[INFO] [1702867875.587127627] [rosbag2_storage]: Opened database '/tmp/tmp_test_dir_SmHn5C/rosbag2_record_and_play_multiple_topics_with_filter_0_sqlite3/rosbag2_record_and_play_multiple_topics_with_filter_0_sqlite3_0.db3' for READ_WRITE.
[WARN] [1702867875.587304449] [rosbag2_cpp]: Environment variable ROS_DISTRO not set, can't store value in bag metadata.
[INFO] [1702867875.591818014] [rosbag2_recorder]: Event publisher thread: Starting
[INFO] [1702867875.591819858] [rosbag2_recorder]: Listening for topics...
[INFO] [1702867875.596126592] [rosbag2_recorder]: Subscribed to topic '/test_topic1'
[INFO] [1702867875.597216102] [rosbag2_recorder]: Subscribed to topic '/test_topic0'
[INFO] [1702867875.598653087] [rosbag2_recorder]: Subscribed to topic '/test_topic'
[INFO] [1702867875.600905696] [rosbag2_recorder]: Subscribed to topic '/rosout'
[INFO] [1702867875.602679117] [rosbag2_recorder]: Subscribed to topic '/events/write_split'
[INFO] [1702867875.602791071] [rosbag2_recorder]: Recording...
terminate called after throwing an instance of 'rosbag2_storage_plugins::SqliteException'
  what():  Error when processing SQL statement. SQLite error (262): database table is locked
/tmp/ws/install_isolated/rosbag2_test_common/include/rosbag2_test_common/rosbag2_test_common/process_execution_helpers_unix.hpp:154: Failure
Expected equality of these values:
  (((status) & 0x7f) == 0)
    Which is: false
```

MichaelOrlov commented:
@emersonknapp Never mind, I have figured it out.
According to https://www2.sqlite.org/cvstrac/wiki?p=DatabaseIsLocked 
Only the first variant is possible 
> Trying to CREATE or DROP a table or index while a SELECT statement is still pending.
Sometimes people think they have finished with a SELECT statement because sqlite3_step() has returned SQLITE_DONE. But the SELECT is not really complete until sqlite3_reset() or sqlite3_finalize() have been called.

It turns out that when we are calling `if (should_split_bagfile(message_timestamp))` here 
https://github.com/ros2/rosbag2/blob/cea3fb027e52b9445266adde9d7629b9957a3c35/rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp#L367-L370
the message_cache consumer thread can call `storage_->write(messages)` at the same time.
There is a race condition and possible situation when two threads will try to execute two different `SELECT` commands on the same database at the same time.

I will try to make protection with mutex and condition_variable on the `rosbag2_storage_sqlite3` layer

morlov-apexai committed:
Protect get_bagfile_size() with database_write_mutex_

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Make get_bagfile_size() lock free to avoid locking double-buffer cache

- Moved db_file_size_ update to the functions which are writing to the
database

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Move test_storage_interfaces.cpp to the rosbag2_tests

- Move test_storage_interfaces.cpp to the rosbag2_tests due to the
cyclic dependencies between rosbag2_storage and storage plugins.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Rename test_storage_interfaces.cpp to the test_rosbag2_storage_api.cpp

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
Update:
It turns out that the idea of synchronizing two threads with mutexes on the `rosbag2_storage_sqlite3 layer` wasn't a good solution.
Since it would cause locking for producer-consumer threads on the upper `SecuantialWriter` layer and discard all benefits of having a double-buffer scheme to split the transport and storage layers.

I made a workaround to avoid locking mutex and doing actual DB requests for the current number of pages. Instead, I moved it to the end of the DB writing functions. This way `get_bagfile_size()` Will return the already pre-evaluated bag file size value. We will do bag size calculation and evaluation per each DB write operation. 
To avoid extra burden on each message write operation when we are dumping the cache buffer to storage I will do a readout for the current number of pages from DB only after finishing dumping a batch of the messages from the cache buffer. 
Also, intermediate filesize evaluation for each `write(message)` call will be made based on the number of bytes written to the DB per each `write(message)` method call.

**Ready for review**

clalancette commented on rosbag2_storage_sqlite3/include/rosbag2_storage_sqlite3/sqlite_storage.hpp:
```diff
@@ -125,13 +125,16 @@ class ROSBAG2_STORAGE_DEFAULT_PLUGINS_PUBLIC SqliteStorage
   RCPPUTILS_TSA_REQUIRES(database_write_mutex_);
   int get_last_rowid();
   int read_db_schema_version();
+  uint64_t get_page_size();
```
This is logically:

```suggestion
  uint64_t get_page_size() const;
```

Though I'm not sure if that will work given the accesses to the database.  Something to consider, though.

MichaelOrlov replied:
It seems ok to make it const. I will add it.

clalancette commented on rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp:
```diff
@@ -323,9 +337,14 @@ void SqliteStorage::write(
 
   for (auto & message : messages) {
     write_locked(message);
+    // Update db_file_size_ with the estimated data size written to the DB
+    // + 3 * sizeof(int64_t) This is for storing timestamp, topic_id and index for messages table
+    db_file_size_ += message->serialized_data->buffer_length + 3 * sizeof(int64_t);
```
Is this necessary?  We are going to throw all of this away after the `commit_transaction` below, so can we just skip this?

MichaelOrlov replied:
Yes. It is required because we have two threads in the rosbag2 recorder. The cache consumer and cache producer. There is possible a situation when we are dumping cache i.e. vector of messages to the disk by the cache consumer and at the same time cache producer requesting up-to-date `db_file_size` that will be needed on the upper layer to decide if bag already exceeds the limit and we need to make a bag split.

clalancette commented on rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp:
```diff
@@ -852,6 +878,36 @@ int SqliteStorage::read_db_schema_version()
 
+uint64_t SqliteStorage::read_total_page_count_locked()
+{
+  auto scope_exit_reset_statement = rcpputils::make_scope_exit(
+    [this]() {
+      page_count_statement_->reset();
+    });
```
Why are we resetting `page_count_statement` after we use it?  I don't know this code very well, but it seems like this statement should just live for the lifetime of the object.

MichaelOrlov replied:
Yeah. I also was confused about this reset when was started implementing this feature and was trying to keep the statement state between invocations. Then I've got some intermittent errors with a message that the database is locked and can't perform other operations on it.  
After some investigation, it turns out that this reset is required for each statement invocation and part of the SQLite API workflow.  
From the `sqlite3.h`
> ** The sqlite3_reset() function is called to reset a [prepared statement]
** object back to its initial state, ready to be re-executed.
** ^Any SQL statement variables that had values bound to them using
** the [sqlite3_bind_blob | sqlite3_bind_*() API] retain their values.
** Use [sqlite3_clear_bindings()] to reset the bindings.
**
** ^The [sqlite3_reset(S)] interface resets the [prepared statement] S
** back to the beginning of its program.
**
** ^If the most recent call to [sqlite3_step(S)] for the
** [prepared statement] S returned [SQLITE_ROW] or [SQLITE_DONE],
** or if [sqlite3_step(S)] has never before been called on S,
** then [sqlite3_reset(S)] returns [SQLITE_OK].
**
** ^If the most recent call to [sqlite3_step(S)] for the
** [prepared statement] S indicated an error, then
** [sqlite3_reset(S)] returns an appropriate [error code].

Also please refer to my findings here https://github.com/ros2/rosbag2/pull/1516#issuecomment-1863342775

clalancette commented on rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp:
```diff
@@ -852,6 +878,36 @@ int SqliteStorage::read_db_schema_version()
+      page_count_statement_->reset();
+    });
+  auto page_count_query_result = page_count_statement_->execute_query<int>();
+  auto page_count_query_result_begin = page_count_query_result.begin();
```
This is just a matter of taste, but it seems unnecessary to assign a variable here.  You are only saving 2 characters while typing :).

MichaelOrlov replied:
Unfortunately not. And in this case this sort of workaround with temporary `page_count_query_result_begin` is required because when we call `page_count_query_result.begin()` we actually calling `statement_->step()` inside and check if we have already called begin() iterator to not allow to call it more than once.
```
    Iterator begin()
    {
      try_access_data();
      return Iterator(statement_, 0);
    }
    void try_access_data()
    {
      if (is_already_accessed_) {
        throw SqliteException("Only one iterator per query result is supported!");
      }
      is_already_accessed_ = true;
    }
```
The rationale for caching `query_result.begin()` is because we need to call it twice (one more time in the return statement at the end of the function) and API allows to call it only once.
One more "skeleton in the SQLite3 wrapper API cabinet"

clalancette commented on rosbag2_storage_sqlite3/include/rosbag2_storage_sqlite3/sqlite_storage.hpp:
```diff
@@ -125,13 +125,16 @@ class ROSBAG2_STORAGE_DEFAULT_PLUGINS_PUBLIC SqliteStorage
   int get_last_rowid();
   int read_db_schema_version();
+  uint64_t get_page_size();
+  uint64_t read_total_page_count_locked() RCPPUTILS_TSA_GUARDED_BY(database_write_mutex_);
```
This is marked as "GUARDED_BY" `database_write_mutex_`, but this isn't a write to the database, is it?  It seems like either the mutex is misnamed, or we need another mutex.

Also, this function is logically "const", though I don't know if it can be that for the same reasons as `get_page_size`.

MichaelOrlov replied:
Yes, probably the name of the mutex is not accurate. The situation is that we can't write and read from the database concurrently. Only one SQL statement is allowed per time. I am guessing that the name originally was chosen as `write` mutex because we didn't have a situation before when it was possible to call DB read and write operations from multiple threads.  
However, I can rename it to the `db_read_write_mutex`.

morlov-apexai committed:
Address review comments

- Rename `database_write_mutex_` to the `db_read_write_mutex_`
- Add const qualifiers for get_page_size() and
read_total_page_count_locked() methods

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
@clalancette Thank you for the review. I've added `const` qualifiers for the functions and renamed mutex.  
Please take a look one more time.

fujitatomoya commented on rosbag2_storage_sqlite3/include/rosbag2_storage_sqlite3/sqlite_storage.hpp:
```diff
@@ -152,11 +155,13 @@ class ROSBAG2_STORAGE_DEFAULT_PLUGINS_PUBLIC SqliteStorage
   // a) database access (this could also be done with FULLMUTEX), but see b)
   // b) topics_ collection - since we could be writing and reading it at the same time
-  std::mutex database_write_mutex_;
+  std::mutex db_read_write_mutex_;
```
this does not block the PR, i thought this is originally going to use `std::shared_mutex` for performance improvement especially read operation, but i only can find write operation for now.

MichaelOrlov replied:
@fujitatomoya I've done some research about SQLite3 multithreaded access.  It turns out that SQLite doesn;t really support multithreaded access to the database.
From the https://www.sqlite.org/threadsafe.html
> SQLite supports three different threading modes:

> Single-thread. In this mode, all mutexes are disabled and SQLite is unsafe to use in more than a single thread at once.

> Multi-thread. In this mode, SQLite can be safely used by multiple threads provided that no single [database connection](https://www.sqlite.org/c3ref/sqlite3.html) nor any object derived from database connection, such as a [prepared statement](https://www.sqlite.org/c3ref/stmt.html), is used in two or more threads at the same time.

> Serialized. In serialized mode, API calls to affect or use any SQLite [database connection](https://www.sqlite.org/c3ref/sqlite3.html) or any object derived from such a database connection can be made safely from multiple threads. The effect on an individual object is the same as if the API calls had all been made in the same order from a single thread. The name "serialized" arises from the fact that SQLite uses mutexes to serialize access to each object.

Therefore as far as I understand correctly even when **Multi-thread** mode is enabled it is not allowed to call more than one prepared statement concurrently from multiple threads. In serialized mode, even worse all requests will be performed sequentially anyway. They will be quied in one thread on the DB level.

In addition to the SQLite3 limitations, we opted out of using DB in the unprotected Single-threaded mode in our `sqlite_wrapper` by explicitly specifying the `SQLITE_OPEN_NOMUTEX` flag during DB open operation.
https://github.com/ros2/rosbag2/blob/195e406194a0bd1901261b8fef180245b70f1a9d/rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_wrapper.cpp#L67-L81
I guess we do this for the sake of performance.

Given everything listed above we have only one choice to use only exclusive mutex.

fujitatomoya replied:
@MichaelOrlov thanks for sharing this, it gives me very good information.

MichaelOrlov replied:
Just to clarify. I think SQLite3 supports multithreaded access in the `Multi-thread` mode but in this case, need to use multiple DB connections. One per each concurrent thread.

fujitatomoya commented on rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp:
```diff
@@ -852,6 +878,36 @@ int SqliteStorage::read_db_schema_version()
+    return 0;  // Trying to call get_page_size when SqliteStorage::open was not called or db
+    // failed to open
+  }
+  SqliteStatement get_page_size_statement = database_->prepare_statement("PRAGMA page_size;");
```
(making sure i understand correctly) i think that is intention not to have `get_page_size_statement_` member variable, instead using stack memory. because `get_page_size` is now called once during opening the sqlite database file, and not expected to be called many time after wards.

MichaelOrlov replied:
@fujitatomoya Yes. your understanding is correct.  
And even if we would have a private member variable for the `get_page_size_statement` we would have to call database_->prepare_statement("PRAGMA page_size;"); and `get_page_size_statement->reset()` each time per `get_page_size` call due to the SQLite3 API limitation. Therefore it doesn't make sense to cache it in the member variable.

fujitatomoya reviewed (approved):
lgtm with green CI

MichaelOrlov commented:
Gist: https://gist.githubusercontent.com/MichaelOrlov/858b61f3724319fdb05b3e8cb28fce6a/raw/2d6648ca88600ddc9ee2e390c25d256f31e2f64b/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_storage rosbag2_storage_sqlite3 rosbag2_tests
TEST args:  --packages-above rosbag2_storage rosbag2_storage_sqlite3 rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/13110
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=20059)](http://ci.ros2.org/job/ci_linux/20059/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=14512)](http://ci.ros2.org/job/ci_linux-aarch64/14512/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=20779)](http://ci.ros2.org/job/ci_windows/20779/)

morlov-apexai committed:
Fix Windows warnings in test_rosbag2_storage_api.cpp

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
Re-run Windows CI build after fixing warnings
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=20785)](http://ci.ros2.org/job/ci_windows/20785/)

MichaelOrlov commented:
@clalancette @emersonknapp The CI is green. However, I need formal approval from those who have write permissions to the rosbag2 repo to be able to merge it.  
Wouldn't you mind to additionally review or approve this PR?

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-for-2024-01-18/35779/1
